pennylane_lightning/core/src/simulators/lightning_kokkos/catalyst/LightningKokkosObsManager.hpp

```diff
@@ -0,0 +1,210 @@
+// Copyright 2023-2024 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <array>
+#include <stdexcept>
+#include <tuple>
+#include <utility>
+
+#include "Types.h"
+#include "Utils.hpp"
+
+#include "ObservablesKokkos.hpp"
+
+namespace Catalyst::Runtime::Simulator {
+
+/**
+ * @brief The LightningKokkosObsManager caches observables of a program at
+ * runtime and maps each one to a const unique index (`int64_t`) in the scope of
+ * the global context manager.
+ */
+template <typename PrecisionT> class LightningKokkosObsManager {
+  private:
+    using VectorStateT =
+        Pennylane::LightningKokkos::StateVectorKokkos<PrecisionT>;
+    using ObservableClassName =
+        Pennylane::Observables::Observable<VectorStateT>;
+    using ObservablePairType =
+        std::pair<std::shared_ptr<ObservableClassName>, ObsType>;
+    std::vector<ObservablePairType> observables_{};
+
+  public:
+    LightningKokkosObsManager() = default;
+    ~LightningKokkosObsManager() = default;
+
+    LightningKokkosObsManager(const LightningKokkosObsManager &) = delete;
+    LightningKokkosObsManager &
+    operator=(const LightningKokkosObsManager &) = delete;
+    LightningKokkosObsManager(LightningKokkosObsManager &&) = delete;
+    LightningKokkosObsManager &operator=(LightningKokkosObsManager &&) = delete;
+
+    /**
+     * @brief A helper function to clear constructed observables in the program.
+     */
+    void clear() { this->observables_.clear(); }
+
+    /**
+     * @brief Check the validity of observable keys.
```

Comment:
We had to add this validator to avoid the misuse of cached observables in the measurement processes either by the compiler or the runtime. In the past, we had runtime bugs related to requesting out-of-bound obs indices in the compiled program due to our observable serialization in the frontend. This could be implemented as an assertion. Here, as `ObsIdType` is a `intptr_t`, we don't really need to check `i >= 0` and this check can be removed. However, `ObsIdType` is an abstract data-type and can be changed by device developers.